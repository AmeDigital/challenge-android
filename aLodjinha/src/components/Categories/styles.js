import { StyleSheet } from 'react-native';
import { RFValue } from 'react-native-responsive-fontsize';
import MaterialDesign from '../../utils/MaterialDesign';

export default StyleSheet.create({
	container: {
		height: 200,
		borderBottomColor: MaterialDesign.colorGreyish,
		borderBottomWidth: 1
	},
	categoryHeader: {
		paddingHorizontal: 15,
		paddingVertical: 15,
		borderBottomWidth: 1,
		borderColor: MaterialDesign.colorGreyish
	},
	categoryTitle: {
		fontSize: RFValue(18),
		color: MaterialDesign.colorDark,
		fontWeight: 'bold'
	},
	categoryContent: {}
});
