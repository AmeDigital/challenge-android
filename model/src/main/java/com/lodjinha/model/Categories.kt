package com.lodjinha.model

data class Categories(val data: List<Category>)
