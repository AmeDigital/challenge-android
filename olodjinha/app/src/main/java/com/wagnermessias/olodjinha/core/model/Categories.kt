package com.wagnermessias.olodjinha.core.model

import com.google.gson.annotations.SerializedName
import java.io.Serializable

data class Categories (
    @SerializedName("data")
    val list: List<Category>
):Serializable