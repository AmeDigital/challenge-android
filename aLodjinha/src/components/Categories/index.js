import React, { useState, useEffect } from 'react';
import { Text, View, ScrollView, Image, TouchableOpacity } from 'react-native';

import api from '../../services/api';

import styles from './styles';
import Loading from '../Loading';
import Category from '../Category';

export default function Categories({ navigation }) {
	const [ categories, setCategories ] = useState([]);

	useEffect(() => {
		(async function loadCategoria() {
			const response = await api.get('/categoria');
			setCategories(response.data.data);
		})();
	}, []);

	function goPageCategories(id, descricao) {
		navigation.navigate('Categories', { id, descricao });
	}

	return (
		<View style={styles.container}>
			<View style={styles.categoryHeader}>
				<Text style={styles.categoryTitle}>Categorias</Text>
			</View>

			{categories.length > 0 ? (
				<ScrollView showsHorizontalScrollIndicator={false} horizontal={true} style={styles.categoryContent}>
					{categories.map((item, index) => (
						<TouchableOpacity key={index} onPress={() => goPageCategories(item.id, item.descricao)}>
							<Category item={item} />
						</TouchableOpacity>
					))}
				</ScrollView>
			) : (
				<Loading />
			)}
		</View>
	);
}
