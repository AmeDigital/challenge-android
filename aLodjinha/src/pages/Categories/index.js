import React, { useState, useEffect } from 'react';
import { FlatList, View, ActivityIndicator } from 'react-native';

import styles from './styles';
import Product from '../../components/Product';
import api from '../../services/api';
import Loading from '../../components/Loading';
import EmotionSad from '../../components/EmotionSad';

function CategoriesScreen({ navigation }) {
	const [ products, setProducts ] = useState([]);
	const [ offset, setOffset ] = useState(0);
	const [ load, setLoad ] = useState(false);
	const [ showEmotion, setShowEmotion ] = useState(false);
	const [ isTotal, setIsTotal ] = useState(false);

	const id = navigation.getParam('id');

	useEffect(() => {
		loadProducts();
	}, []);

	async function loadProducts() {
		if (load) return;

		if (isTotal) return;

		setLoad(true);

		try {
			const response = await api.get(`/produto?limit=20&offset=${offset}&categoriaId=${id}`);

			setProducts([ ...products, ...response.data.data ]);

			if (response.data.data.length === 0) {
				setShowEmotion(true);
				return;
			}

			/**Logica para limitar o tamanho da lista */
			if (response.data.data.length < 20) {
				setIsTotal(true);
			}

			setOffset(offset + 2);
			setLoad(false);
		} catch (e) {
			console.error(e);
		}
	}

	function funcPage(id, descricao) {
		navigation.navigate('DescriptionProduct', { id, descricao });
	}

	function listCategories() {
		return (
			<View style={styles.container}>
				{products.length > 0 ? (
					<FlatList
						data={products}
						renderItem={({ item }) => <Product key={item.id} funcPage={funcPage} item={item} />}
						keyExtractor={(item) => String(item.id + Math.floor(Math.random() * 1985419652))}
						onEndReached={loadProducts}
						onEndReachedThreshold={0.1}
						ListFooterComponent={
							load && <ActivityIndicator size="large" color="#5e2a84" style={{ margin: 20 }} />
						}
					/>
				) : (
					<Loading />
				)}
			</View>
		);
	}

	return showEmotion ? <EmotionSad /> : listCategories();
}

CategoriesScreen.navigationOptions = ({ navigation }) => {
	return {
		title: navigation.getParam('descricao')
	};
};

export default CategoriesScreen;
