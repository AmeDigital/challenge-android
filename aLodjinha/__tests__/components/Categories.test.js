import React from 'react';

import renderer from 'react-test-renderer';

import Categories from '../../src/components/Categories';

describe('Component Categories', () => {
	test('should render without erros', () => {
		const tree = renderer.create(<Categories />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
