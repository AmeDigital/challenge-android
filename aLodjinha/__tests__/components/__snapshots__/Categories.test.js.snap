// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Categories should render without erros 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#aaaaaa",
      "borderBottomWidth": 1,
      "height": 200,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#aaaaaa",
        "paddingHorizontal": 15,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#2d3142",
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      Categorias
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#5e2a84"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
